File: Resoluções-das-Listas/Lista-1/Lista-1-Problema-3.cpp
#include <iostream>
#include <fstream>
#include <cmath>

using namespace std;

// Função f(x) = e^(-x) * cos(x)
double f(double x)
{
    return exp(-x) * cos(x);
}

// Derivada numérica progressiva
double derivada_progressiva(double x, double h)
{
    return (f(x + h) - f(x)) / h;
}

// Derivada numérica centrada
double derivada_centrada(double x, double h)
{
    return (f(x + h) - f(x - h)) / (2 * h);
}

int main()
{
    double x_min, x_max, h;
    ofstream arquivo("resultados.txt");

    // Solicita os valores de x_min, x_max e h
    cout << "Digite o valor de x_min: ";
    cin >> x_min;
    cout << "Digite o valor de x_max: ";
    cin >> x_max;
    cout << "Digite o valor de h: ";
    cin >> h;

    // Verifica se o arquivo foi aberto corretamente
    if (!arquivo.is_open())
    {
        cerr << "Erro ao abrir o arquivo!" << endl;
        return 1;
    }

    // Escreve os resultados no arquivo
    arquivo << "x\tProgressiva\tCentrada" << endl;
    for (double x = x_min; x <= x_max; x += h)
    {
        double dp = derivada_progressiva(x, h);
        double dc = derivada_centrada(x, h);
        arquivo << x << "\t" << dp << "\t" << dc << endl;
    }

    arquivo.close();
    cout << "Resultados escritos em 'resultados.txt'" << endl;

    return 0;
}

File: Resoluções-das-Listas/Lista-1/Lista-1-Problema-3.py
import math

# Função f(x) = e^(-x) * cos(x)


def f(x):
    return math.exp(-x) * math.cos(x)

# Derivada numérica progressiva


def derivada_progressiva(x, h):
    return (f(x + h) - f(x)) / h

# Derivada numérica centrada


def derivada_centrada(x, h):
    return (f(x + h) - f(x - h)) / (2 * h)


def main():
    # Solicita os valores de x_min, x_max e h
    x_min = float(input("Digite o valor de x_min: "))
    x_max = float(input("Digite o valor de x_max: "))
    h = float(input("Digite o valor de h: "))

    # Abre o arquivo para escrita
    with open("resultados.txt", "w") as arquivo:
        # Escreve os resultados no arquivo
        arquivo.write("x\tProgressiva\tCentrada\n")
        x = x_min
        while x <= x_max:
            dp = derivada_progressiva(x, h)
            dc = derivada_centrada(x, h)
            arquivo.write(f"{x}\t{dp}\t{dc}\n")
            x += h

    print("Resultados escritos em 'resultados.txt'")
